commit 83c4e43722a2c8a8438b8d165047720fd36aaea4
Author: Al Viro <viro@parcelfarce.linux.theplanet.co.uk>
Date:   Fri Aug 19 15:56:37 2005 -0700

    [SPARC]: Fix weak aliases
    
    sparc_ksyms.c used to declare weak alias to several gcc intrinsics.  It
    doesn't work with gcc4 anymore - it wants a declaration for the thing
    we are aliasing to and that's not going to happen for something like
    .mul, etc.  Replaced with direct injection of weak alias on the assembler
    level - .weak <alias> followed by <alias> = <aliased>; that works on all
    gcc versions.
    
    Signed-off-by: Al Viro <viro@parcelfarce.linux.theplanet.co.uk>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/arch/sparc/kernel/sparc_ksyms.c b/arch/sparc/kernel/sparc_ksyms.c
index 1bd430d0ca06..8faa8dc4de43 100644
--- a/arch/sparc/kernel/sparc_ksyms.c
+++ b/arch/sparc/kernel/sparc_ksyms.c
@@ -98,8 +98,9 @@ extern void ___rw_write_enter(void);
  * The module references will be fixed up by module_frob_arch_sections.
  */
 #define DOT_ALIAS2(__ret, __x, __arg1, __arg2) \
-	extern __ret __x(__arg1, __arg2) \
-	             __attribute__((weak, alias("." # __x)));
+	extern __ret __x(__arg1, __arg2); \
+	asm(".weak " #__x);\
+	asm(#__x "=." #__x);
 
 DOT_ALIAS2(int, div, int, int)
 DOT_ALIAS2(int, mul, int, int)